File: 004 链表的基本操作实现/main.c
#include <stdio.h>
#include <stdlib.h>

typedef struct Node {
    int value;
    struct Node *next;
} Node, LinkList;

void * createList() {
    LinkList *list = (LinkList *)malloc(sizeof(LinkList));
    list->next = NULL;
    list->value = 0;
    return list;
}

void insert(LinkList *list, int idx, int value) {
    if (idx < 0) {
        printf("插入失败\n");
    } else {
        Node *p = list;
        int i = 0;
        while (p != NULL && i < idx) {
            p = p->next;
            i++;
        }
        
        if (p != NULL && i == idx) {
            Node *n = (Node *)malloc(sizeof(Node));
            n->value = value;
            n->next = p->next;
            p->next = n;
            list->value++;
        } else {
            printf("插入失败\n");
        }
    }
}

void append(LinkList *list, int value) {
    insert(list, list->value, value);
}

void delete(LinkList *list, int idx) {
    if (idx < 0) {
        printf("删除失败\n");
    } else {
        Node *p = list;
        int i = 0;
        while (p != NULL && i < idx) {
            p = p->next;
            i++;
        }
        Node *next = p->next;
        if (next != NULL && i == idx) {
            p->next = next->next;
            free(next);
            list->value--;
        } else {
            printf("删除失败\n");
        }
    }
}

void update(LinkList *list, int idx, int value) {
    if (idx < 0) {
        printf("更新失败\n");
    } else {
        Node *p = list->next;
        int i = 0;
        while (p != NULL && i < idx) {
            p = p->next;
            i++;
        }
        
        if (p != NULL && i == idx) {
            p->value = value;
        } else {
            printf("更新失败\n");
        }
    }
}

void getValue(LinkList *list, int idx, int *value) {
    if (idx < 0) {
        printf("获取失败\n");
    } else {
        Node *p = list->next;
        int i = 0;
        while (p != NULL && i < idx) {
            p = p->next;
            i++;
        }
        if (p != NULL && i == idx) {
            *value = p->value;
        } else {
            printf("获取失败\n");
        }
    }
}

void getIndex(LinkList *list, int value, int *idx) {
    Node *p = list->next;
    int i = 0;
    while (p != NULL) {
        if (p->value == value) {
            *idx = i;
            break;
        }
        p = p->next;
        i++;
    }
}

void length(LinkList *list) {
    return list->value;
}

void clean(LinkList *list) {
    Node *p = list->next;
    list->next = NULL;
    list->value = 0;
    
    while (p != NULL) {
        Node *next = p->next;
        free(p);
        p = next;
    }
}

void destroy(LinkList *list) {
    clean(list);
    free(list);
}

void traverse(LinkList *list) {
    Node *p = list->next;
    while (p != NULL) {
        printf("%d ", p->value);
        p = p->next;
    }
    printf("\n");
}

int main() {
    LinkList *list = createList();
    traverse(list);
    
    for (int i = 0; i < 10; i++) {
        int value = random() % 20;
        printf("%d ", value);
        append(list, value);
    }
    printf("\n");
    traverse(list);
    
    insert(list, 5, 100);
    traverse(list);
    
    delete(list, 10);
    traverse(list);
    
    update(list, 5, 200);
    traverse(list);
    
    int result = -1;
    getValue(list, 7, &result);
    printf("%d \n", result);

    int idx = -1;
    getIndex(list, 6, &idx);
    printf("%d \n", idx);
    
    
    return 0;
}

File: 004 链表的基本操作实现/main.swift
import Foundation

class LinkList<T: Equatable> {
    class Node<T> {
        var value: T? = nil
        var next: Node<T>? = nil
    }

    private let header = Node<T>()
    private var length = 0
    
    func insert(_ value: T, at idx: UInt) {
        var p: Node<T>? = header
        var i: UInt = 0
        while p != nil && i < idx {
            p = p?.next
            i += 1
        }
        
        if p != nil && i == idx {
            let n = Node<T>()
            n.next = p?.next
            n.value = value
            
            p?.next = n
            length += 1
        } else {
            print("插入失败")
        }
    }
    
    func append(_ value: T) {
        insert(value, at: UInt(length))
    }
    
    func delete(at idx: UInt) {
        var p: Node<T>? = header
        var i: UInt = 0
        while p != nil && i < idx {
            p = p?.next
            i += 1
        }
        let next = p?.next
        if next != nil && i == idx {
            p?.next = next?.next
            length -= 1
        } else {
            print("删除失败")
        }
    }
    
    func update(value: T, at idx: UInt) {
        var p: Node<T>? = header.next
        var i: UInt = 0
        while p != nil && i < idx {
            p = p?.next
            i += 1
        }
        if p != nil && i == idx {
            p?.value = value
        } else {
            print("更新失败")
        }
    }
    
    func getValue(of idx: UInt) -> T? {
        var p: Node<T>? = header.next
        var i: UInt = 0
        while p != nil && i < idx {
            p = p?.next
            i += 1
        }
        if p != nil && i == idx {
            return p?.value
        } else {
            return nil
        }
    }
    
    func getIndex(of value: T) -> UInt? {
        var p: Node<T>? = header.next
        var i: UInt = 0
        while p != nil {
            if p?.value == value {
                return i
            }
            p = p?.next
            i += 1
        }
        return nil
    }
    
    func clear() {
        header.next = nil
        length = 0
    }
    
    func traverse() {
        var p = header.next
        while p != nil {
            print(p!.value!, terminator: " ")
            p = p?.next
        }
        print()
    }
}

let list = LinkList<Int>()
list.traverse()

for _ in 0..<10 {
    let value = Int.random(in: 1...20)
    print(value, terminator: " ")
    list.append(value)
}
print()
list.traverse()

list.delete(at: 3)
list.traverse()

list.update(value: 100, at: 7)
list.traverse()

print(list.getValue(of: 7))

print(list.getIndex(of: 100))
